### PR TITLE
Enable PHPCS rules for PHP 8.0

### PR DESCRIPTION
**Pull request by derrabus**
This will add some types, especially to the test suite. ~~Let's see which tests fail and fix the issues on lower branches.~~

### CODE REVIEW

File: lib/Doctrine/ORM/Cache/Persister/Entity/NonStrictReadWriteCachedEntityPersister.php
Comment (SenseException) [inline lib/Doctrine/ORM/Cache/Persister/Entity/NonStrictReadWriteCachedEntityPersister.php]: Can `$entity` get an `object` type hint? If not, then `$entity::class` shouldn't be used.

File: lib/Doctrine/ORM/Id/AssignedGenerator.php
Comment (SenseException) [inline lib/Doctrine/ORM/Id/AssignedGenerator.php]: Can `$entity` be `null`? If not, then `$entity::class` shouldn't be used.

File: lib/Doctrine/ORM/Mapping/DefaultEntityListenerResolver.php
Comment (SenseException) [inline lib/Doctrine/ORM/Mapping/DefaultEntityListenerResolver.php]: Same as mentioned before. I won't comment the other lines with `$variable::class`

**Review by SenseException — COMMENTED** [review]
This PR feels like it should be a follow up to a PR that introduces more type hints in all methods.

File: lib/Doctrine/ORM/Cache/Persister/Entity/NonStrictReadWriteCachedEntityPersister.php
Comment (derrabus) [reply lib/Doctrine/ORM/Cache/Persister/Entity/NonStrictReadWriteCachedEntityPersister.php]: We shouldn't call `get_class()` either then.

I mean, we could deactivate that rule if we don't feel comfortable about it. But the `::class` syntax should be as (un)safe as a `get_class()` call.

**Comment by derrabus** [comment]
Thank you for the review @SenseException. I think, I should've been more clear about how this PR was generated. All I did was raising the PHP language level for CodeSniffer. All other changes are automated with `phpcbf`. This is how our Doctrine coding standard wants a PHP 8 codebase to look like. 🙂 

> This PR feels like it should be a follow up to a PR that introduces more type hints in all methods.

Yes, and that's my plan. Applying types everywhere fails miserably (see #9270), mainly because our doc blocks are not as accurate as they need to be for automating that change.

Instead, I try to add types iteratively (see #9292) and if I find inaccurate typings while doing so, I backport the type fixes to 2.11.x (see #9273). Feel free to join that effort. ORM 3.0 is a good oppertunity to apply native types throughout the codebase.

File: lib/Doctrine/ORM/Id/AssignedGenerator.php
Comment (derrabus) [reply lib/Doctrine/ORM/Id/AssignedGenerator.php]: Same as above, `get_class()` fails on `null` as well.

`AbstractIdGenerator::generate()` declares the parameter as nullable, but that doesn't feel right. Maybe we should evaluate tighten the parameter type here, but I'd consider this out of scope for this PR.

**Comment by SenseException** [comment]
I like the change itself because a `null` value would consistently result in a fatal error in all supported PHP versions. I'm more comfortable when at least the involved methods can have a type hint, because if the phpdocs aren't accurate now, it might be more painful for those who need to migrate. Do I need to review one of the drafts in case of introducing types? Maybe partially introducing types where it works?

**Comment by derrabus** [comment]
> I like the change itself because a `null` value would consistently result in a fatal error in all supported PHP versions.

"All supported PHP versions" on the 3.0.x branch are 8.0 and 8.1. Given `$foo = null`, both `get_class($foo)` and `$foo::class` would produce a `TypeError` on these versions. The error message changes, that's all.

* https://3v4l.org/5ZKZ1
* https://3v4l.org/vdLJD

Yes, we should get the types straight, but let's do this in follow-ups. This PR sets up our tooling for the new language level. Once it's merged we can work on adding native type declarations where it makes sense.

File: phpcs.xml.dist
Comment (derrabus) [inline phpcs.xml.dist:29]: I've deactivated the rule for now because the discussion around it seems to block the merge. I can activate it again in a follow-up.

**Comment by SenseException** [comment]
Okay, once the build is green, I'll :+1:

**Review by SenseException — APPROVED** [review]

**Pull request merged**
